File: src/bme/creditcalc/ui/SemesterTable.java
package bme.creditcalc.ui;

import bme.creditcalc.model.Semester;

import javax.swing.table.AbstractTableModel;

public class SemesterTable extends AbstractTableModel {
    Semester semester;

    public Semester getSemester() {
        return semester;
    }

    public void setModel(Semester s){
        semester = s;
        fireTableDataChanged();
    }

    @Override
    public int getRowCount() {
        return semester.getSubjects().size();
    }

    @Override
    public int getColumnCount() {
        return 4;
    }

    @Override
    public String getColumnName(int column) {
        switch (column){
            case 0:
                return "Name";
            case 1:
                return "Credit";
            case 2:
                return "Grade";
            case 3:
                return "Minta";
        }
        return super.getColumnName(column);
    }

    @Override
    public Object getValueAt(int rowIndex, int columnIndex) {
        switch (columnIndex){
            case 0:
                return semester.getSubjects().get(rowIndex).getName();
            case 1:
                return semester.getSubjects().get(rowIndex).getCredit();
            case 2:
                return semester.getSubjects().get(rowIndex).getGrade();
            case 3:
                return semester.getSubjects().get(rowIndex).getMinta();
            default:
                return null;
        }
    }


    @Override
    public boolean isCellEditable(int rowIndex, int columnIndex) {
        return true;
        /*if(columnIndex > 1){
            return true;
            //return !subjects.get(rowIndex).isFinalized();
        }
        return false;*/
    }

    @Override
    public Class<?> getColumnClass(int columnIndex) {
        switch (columnIndex){
            case 0:
                return String.class;
            case 1:
                return Double.class;
            case 2:
                return Integer.class;
            case 3:
                return Boolean.class;
        }
        return Object.class;
    }

    @Override
    public void setValueAt(Object aValue, int rowIndex, int columnIndex) {
        switch (columnIndex){
            case 0:
                semester.getSubjects().get(rowIndex).setName((String) aValue);
                break;
            case 1:
                semester.getSubjects().get(rowIndex).setCredit((Double) aValue);
                break;
            case 2:
                semester.getSubjects().get(rowIndex).setGrade((Integer) aValue);
                this.fireTableRowsUpdated(rowIndex, rowIndex);
                break;
            case 3:
                semester.getSubjects().get(rowIndex).setMinta((Boolean) aValue);
                this.fireTableRowsUpdated(rowIndex, rowIndex);
                break;
            default:
                break;
        }
    }
}

File: src/main/java/bme/creditcalc/ui/SemesterTable.kt
package bme.creditcalc.ui

import bme.creditcalc.model.Semester
import javax.swing.table.AbstractTableModel

class SemesterTable : AbstractTableModel() {
    var semester: Semester? = null
    fun setModel(s: Semester?) {
        semester = s
        fireTableDataChanged()
    }

    override fun getRowCount(): Int {
        return semester?.subjects?.size ?: 0
    }

    override fun getColumnCount(): Int {
        return 4
    }

    override fun getColumnName(column: Int): String {
        when (column) {
            0 -> return "Name"
            1 -> return "Credit"
            2 -> return "Grade"
            3 -> return "Minta"
        }
        return super.getColumnName(column)
    }

    override fun getValueAt(rowIndex: Int, columnIndex: Int): Any {
        if(semester == null){
            return ""
        }
        return when (columnIndex) {
            0 -> semester!!.subjects[rowIndex].name
            1 -> semester!!.subjects[rowIndex].credit
            2 -> semester!!.subjects[rowIndex].grade
            3 -> semester!!.subjects[rowIndex].minta
            else -> semester!!.subjects[rowIndex].name
        }
    }

    override fun isCellEditable(rowIndex: Int, columnIndex: Int): Boolean {
        return true
    }

    override fun getColumnClass(columnIndex: Int): Class<*> {
        return when (columnIndex) {
            0 -> String::class.javaObjectType
            1 -> Double::class.javaObjectType
            2 -> Int::class.javaObjectType
            3 -> Boolean::class.javaObjectType
            else -> Any::class.javaObjectType
        }
    }

    override fun setValueAt(aValue: Any, rowIndex: Int, columnIndex: Int) {
        if(semester == null){
            return
        }
        when (columnIndex) {
            0 -> semester!!.subjects[rowIndex].name = aValue as String
            1 -> semester!!.subjects[rowIndex].credit = aValue as Double
            2 -> {
                semester!!.subjects[rowIndex].grade = aValue as Int
                fireTableRowsUpdated(rowIndex, rowIndex)
            }
            3 -> {
                semester!!.subjects[rowIndex].minta = aValue as Boolean
                fireTableRowsUpdated(rowIndex, rowIndex)
            }
            else -> {}
        }
    }
}